contracts/ERC1238/ERC1238.sol

```diff
@@ -107,14 +123,138 @@ contract ERC1238 is IERC1238 {
+     */
+    function _mintToEOA(
+        address to,
+        uint256 id,
+        uint256 amount,
+        uint8 v,
+        bytes32 r,
+        bytes32 s,
+        bytes memory data
+    ) internal virtual {
+        bytes32 messageHash = _getMintApprovalMessageHash(to, id, amount);
+        _verifyMintingApproval(to, messageHash, v, r, s);
+
+        _mint(to, id, amount, data);
+    }
+
+    /**
+     * @dev [Batched] version of {_mintToContract}. A batch specifies an array of token `id` and
+     * the amount of tokens for each.
+     *
+     * Requirements:
+     * - `to` must be a smart contract and must implement {IERC1238Receiver-onERC1238BatchMint} and return the
+     * acceptance magic value.
+     * - `ids` and `amounts` must have the same length.
+     *
+     * Emits a {MintBatch} event.
+     */
+    function _mintBatchToContract(
+        address to,
+        uint256[] memory ids,
+        uint256[] memory amounts,
+        bytes memory data
+    ) internal virtual {
+        require(to.isContract(), "ERC1238: Recipient is not a contract");
+
+        _mintBatch(to, ids, amounts, data);
+
+        _doSafeBatchMintAcceptanceCheck(msg.sender, to, ids, amounts, data);
+    }
+
+    /**
+     * @dev [Batched] version of {_mintToEOA}. A batch specifies an array of token `id` and
+     * the amount of tokens for each.
+     *
+     * Requirements:
+     * - `v`, `r` and `s` must be a EIP712 signature from `to` as defined by ERC1238Approval to
+     * approve the batch minting transaction.
+     *
+     * Emits a {MintBatch} event.
+     */
+    function _mintBatchToEOA(
+        address to,
+        uint256[] memory ids,
+        uint256[] memory amounts,
+        uint8 v,
+        bytes32 r,
+        bytes32 s,
+        bytes memory data
+    ) internal virtual {
+        bytes32 messageHash = _getMintBatchApprovalMessageHash(to, ids, amounts);
+        _verifyMintingApproval(to, messageHash, v, r, s);
+
+        _mintBatch(to, ids, amounts, data);
+    }
+
+    /**
+     * @dev Mints a bundle, which can be viewed as minting several batches
+     * to an array of addresses in one transaction.
+     *
+     * Requirements:
+     * - `to` can be a combination of smart contract addresses and EOAs.
+     * - If `to` is not a contract, an EIP712 signature from `to` as defined by ERC1238Approval
+     * must be passed at the right index in `data`.
+     *
+     * Emits multiple {MintBatch} events.
+     */
+    function _mintBundle(
+        address[] memory to,
+        uint256[][] memory ids,
+        uint256[][] memory amounts,
+        bytes[] memory data
+    ) internal virtual {
+        for (uint256 i = 0; i < to.length; i++) {
+            if (to[i].isContract()) {
+                _mintBatchToContract(to[i], ids[i], amounts[i], data[i]);
+            } else {
+                (bytes32 r, bytes32 s, uint8 v) = splitSignature(data[i]);
+                _mintBatchToEOA(to[i], ids[i], amounts[i], v, r, s, data[i]);
```

Comment:
I wonder if we should slice away the first 130-bytes if we are still passing the data object through? This way we can remove the signature section that we've already computed and allow the function to just handle any additional data.

Reply:
Let's create an issue, at first I thought it can just be ignored but I agree I don't like the inconsistency it creates. Sound like slicing away is worth it for implementers to spare them the hassle of distinguishing if data contains a signature or not. 
Other solutions is to have another array dedicated to signatures...

Reply:
Yep, let's continue discussion on an issue and make it an improvement and not delay this PR.